### PR TITLE
Cache last commit when pushing for big repository

### PR DESCRIPTION
**Pull request by lunny**
This PR will cache the last commit ids for the home page of big repositories entries when a new commit pushed. Then when you visit the home page first time after a new commit pushed, it should be still fast enough.

~Still WIP, wait for a queue since it may spent many times.~

~Blocked by #10133~

### CODE REVIEW

**Comment by bagasme** [comment]
How many commits a repo can be considered big?

**Comment by lunny** [comment]
@bagasme there is a config, default is 1000 commits.

File: services/repository/push.go [hunk 1]
```diff
@@ -224,9 +245,27 @@ func pushUpdates(optsList []*PushUpdateOptions) error {
...
+				}
 
-				go pull_service.AddTestPullRequestTask(pusher, repo.ID, branch, true, opts.OldCommitID, opts.NewCommitID)
+				// Cache for big repository
+				if err := repo_module.CacheRef(repo, gitRepo, opts.RefFullName); err != nil {
+					log.Error("repo_module.CacheRef %s/%s failed: %v", repo.ID, branch, err)
+				}
+
+				log.Trace("TriggerTask '%s/%s' by %s", repo.Name, branch, pusher.Name)
+				pull_service.AddTestPullRequestTask(pusher, repo.ID, branch, true, opts.OldCommitID, opts.NewCommitID)
```
Comment (6543) [inline services/repository/push.go]: can we move this before commit cache?

**Comment by 6543** [comment]
@lunny good idear, we only should take care the easy jobs done while pushing should be first (trigger webhooks, calc commit count cache, ..) and the slowest tasks should be add the end

**Comment by lunny** [comment]
It needs more refactor here. I will send other PRs.

**Comment by 6543** [comment]
CI Failed

**Review by techknowlogick — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Pull request merged**